File: 0034-find-first-and-last-position-of-element-in-sorted-array/0034-find-first-and-last-position-of-element-in-sorted-array.cpp
class Solution {
public:
    vector<int> searchRange(vector<int>& nums, int target) {
        int first = -1, last = -1;
        int l = 0, r = nums.size() - 1;
        while (l < r) {
            int mid = (l + r) / 2;
            if (nums[mid] < target) {
                l = mid + 1;
            } else {
                r = mid;
            }
        }
        if (r >= 0 && nums[r] == target) {
            first = r;
        }

        l = 0, r = nums.size() - 1;
        while (l <= r) {
            int mid = (l + r) / 2;
            if (nums[mid] > target) {
                r = mid - 1;
            } else {
                l = mid + 1;
            }
        }
        if (r >= 0 && nums[r] == target) {
            last = r;
        }

        return {first, last};
    }
};

File: 0034-find-first-and-last-position-of-element-in-sorted-array/0034-find-first-and-last-position-of-element-in-sorted-array.py
class Solution:
    def searchRange(self, nums: List[int], target: int) -> List[int]:
        x, y = -1, -1
        n = len(nums)
        # ====================

        # lower_bound
        i, j = 0, n - 1
        while i <= j:
            mid = (i + j) // 2
            if nums[mid] < target:
                i = mid + 1
            else:
                j = mid - 1
        if i < n and nums[i] == target:
            x = i
        # print(x)

        # ====================

        # upper_bound
        i, j = 0, n - 1
        while i <= j:
            mid = (i + j) // 2
            if nums[mid] > target:
                j = mid - 1
            else:
                i = mid + 1
        if i - 1 < n and i - 1 >= 0 and nums[i - 1] == target:
            y = i - 1
        # print(y)

        # ====================

        return [-1, -1] if x == -1 or y == -1 else [x, y]
